### ISSUE TITLE
[要望] メッシュ裏面を張りたい

### ISSUE THREAD

**Issue by Narazaka** [kurotu/VRCQuestTools#26]
Androidで使用可能なシェーダーでは両面描画が無理なため、ポリゴンをNDMF Mantisとかで減らした上で法線反転したメッシュを生成したい（主にスカート表現などのため）

というjust ideaです

これ系どちらかといえばAAOとかのプロジェクトの方がノウハウあるかもですが、ジャンルとしてはVQTだなと思ったのでとりあえず

**Comment by kurotu** [comment]
思いつきレベルですけどこれをNDMF化するのが手っ取り早いかなと思っています。まだ何も確認してませんが。

https://booth.pm/ja/items/5645609

メッシュが反転したまま販売されている衣装がたまにあったりするので、それを直すために単独のツールでもいいのかなと思ったりしますが、PCだとメッシュの向きはあまり気にしない感じですかね。

**Comment by Narazaka** [comment]
そんな衣装あるんだ
デフォルトセットアップ以外にシェーダーパラメーターを弄った場合などにたまに発見されるくらいな気がしますね……

**Comment by kurotu** [comment]
完全と言える作り込みに時間がかかりそうだったため v2.6.0 で実験的機能として追加しました。
実際のユースケースで気付いた点があればフィードバックをいただきたいです。

**Comment by Narazaka** [comment]
個人的に所持している衣装の傾向でいうと、これが必要となるのが比較的ロングなスカートでワンピース様になりやすい？ため、1オブジェクト全部のメッシュを反転させるよりかはマスクかけたとこだけ反転したいケースは多いかもと思いました。
ただオブジェクトを複製してAAOのマスクでいらんとこけしてこれをかけてから結合するという風に現時点でも組めなくはないので、回避は可能だと思います（面倒＆エッジ部分が異なるかもではあるけど）。

**Comment by Narazaka** [comment]
いや、Mantis LODとかがかかる前提だとAAOで削減したメッシュとそうでないものの結果が変わる可能性があるので代替できるとはいえない……？

**Comment by kurotu** [comment]
マスクでの反転はぜひ入れたいところですがデシメーション (Mantis LOD, lilNDMFMeshSimplifier) との組み合わせというか実行順の調整は実際困りそうです。実際作ってみないと何とも言えないところですが。

反転の意図を反映しやすいという点ではデシメーション前の方がマシですかね？

### デシメーション前に反転
#### Pros
- 素直に反転処理できる
#### Cons
- 反転でポリゴン数が増えた後にポリゴン削減するので、最終的な形状を予測しづらい
(現状は頂点数を増やしていないので杞憂な気はしつつ、真面目に法線を処理し始めると影響あるかも？)

### デシメーション後に反転
#### Pros
- 最終的な形状やポリゴン数を予測しやすい
#### Cons
- UVが変わってしまいマスクの意図通り反転できない可能性がある

**Comment by Narazaka** [comment]
AAOとかが前パスっぽいので前っぽくはありつつ、理想的には前パス後パスどっちもあって選べるとか……

**Comment by kurotu** [comment]
v2.7.0でマスクテクスチャに対応しました。
意図通り反転させることを優先し、デシメーション前に処理するように変更しています。

Standard Liteへの対応時に法線の処理を確認すればMesh Flipperの機能としては一区切りかなと考えています。

マスクテクスチャの編集機能はたぶんつけないと思います。ちょっと大変すぎるので...
(AAOの実装流用できないかとかも一瞬考えましたが、そう簡単にはいかなそう)

**Comment by Narazaka** [comment]
まあいまどきのQuest対応においてAAOを入れないことはあり得ないと思うので、必要ならAAOのやつ使ってくれスタンスでいいような気はします

**Comment by Narazaka** [comment]
NDMF Mantisのデシメーションの設定を変えずにMesh Flipperの適用有無を切り替えたところ、謎にぎざぎざのメッシュになりやすい感じでした。
特定メッシュだけでなくスカート等別メッシュでも同様の傾向がありました。
デシメーションコンポーネントのプレビューではなかなか予測不可能な感じの結果になるので、後パスもほしいかも……。

![Image](https://github.com/user-attachments/assets/896ec494-3ba6-4cc8-9e64-244e12f7ee69)

![Image](https://github.com/user-attachments/assets/63f55345-d73a-4669-ad22-c742a507c9f8)

**Comment by kurotu** [comment]
デシメーションの予測できなさが結構なものだったので、v2.9.0ではフェーズ設定を入れつつデフォルトをOptimizing Phaseで動作するように変更しました。
また、結果を確認できるように実験的にですがNDMF Previewの対応を入れました。Mobileシェーダーにしていないと変なことになるので現状はデフォルトで無効です。

**Comment by kurotu** [comment]
Toon Standardでは裏面描画ができるようになってコンポーネントとしての利用場面は減ってきそうです。また、Standard Liteとの組み合わせで法線が変な感じにはなっていなかったので大きな変更も今後はあまり必要なさそうです。
上記の点で、次回 v2.10.0 では実験的フラグを外してこのIssueとしては一旦closeにしようと思っています。

**Comment by kurotu** [comment]
v2.10.0で実験的コンポーネントのフラグを外しました。
何かあれば、通常通り個別にIssueを立ててください。
